File: 2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        if (head == null || head.next == null || head.next.next == null) {
            return new int[]{-1, -1};
        }
        
        int firstCriticalIndex = -1;
        int lastCriticalIndex = -1;
        int minDistance = Integer.MAX_VALUE;
        int index = 1;
        
        ListNode prev = head;
        ListNode curr = head.next;
        
        while (curr.next != null) {
            if ((curr.val > prev.val && curr.val > curr.next.val) ||
                (curr.val < prev.val && curr.val < curr.next.val)) {
                
                if (firstCriticalIndex == -1) {
                    firstCriticalIndex = index;
                } else {
                    minDistance = Math.min(minDistance, index - lastCriticalIndex);
                }
                lastCriticalIndex = index;
            }
            prev = curr;
            curr = curr.next;
            index++;
        }
        
        if (firstCriticalIndex == lastCriticalIndex) {
            return new int[]{-1, -1};
        }
        
        int maxDistance = lastCriticalIndex - firstCriticalIndex;
        return new int[]{minDistance, maxDistance};
    }
}

File: 2058. Find the Minimum and Maximum Number of Nodes Between Critical Points.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn nodes_between_critical_points(head: Option<Box<ListNode>>) -> Vec<i32> {
        let mut prev = match head.as_ref() {
            Some(val) => val,
            None => return vec![-1, 1]
        };
        let mut curr = prev.next.as_ref();

        let mut i = 0;
        let mut first = None;
        let mut last = None;
        let mut min_distance = None;
        while let Some(c) = curr {
            i += 1;
            let Some(next) = c.next.as_ref() else {
                break
            };
            if (c.val > prev.val && c.val > next.val) || 
                (c.val < prev.val && c.val < next.val) {
                first.get_or_insert(i);
                if let Some(last) = last {
                    let distance = Some(i - last);
                    min_distance = min_distance.or(distance).min(distance);
                }
                last = Some(i);
            }
            prev = c;
            curr = Some(next);
        }
        
        match (min_distance, last, first) {
            (Some(min_distance), Some(last), Some(first)) => vec!(min_distance, last-first),
            _ => vec![-1,-1]
        }
    }
}
